File: leetcode/7.py
#!/usr/bin/env python

"""
Reverse Integer
Reverse digits of an integer.

Example1: x = 123, return 321
Example2: x = -123, return -321
"""


class Solution(object):
    def reverse(self, x):
        """
        :type x: int
        :rtype: int
        """

        int_max = 2147483647 / 10

        negative = False
        if x < 0:
            negative = True
            x = -x

        reversed = 0
        while x != 0:
            if reversed > int_max:
                # overflow
                return 0

            reminder = x % 10
            x = x / 10
            reversed = reversed * 10 + reminder

        if negative:
            reversed = -reversed

        return reversed

File: leetcode/7.cpp
#include <iostream>

using namespace std;

class Solution {
public:
    int reverse(int x) {
        int max_bound = INT_MAX / 10;
        bool negative = 0;
        if (x < 0)
        {
            if (x == INT_MIN)
            {
                return 0;
            }

            negative = 1;
            x = -x;
        }

        int reversed = 0;
        while (x != 0)
        {
            if (reversed > max_bound)
            {
                return 0;
            }

            int reminder = x % 10;
            x /= 10;
            reversed = reversed * 10 + reminder;
        }

        if (negative == 1)
        {
            reversed = -reversed;
        }

        return reversed;
    }
};
